Add compatible logging when parsing a compatible field

pgomulka opened the pull request:
A #68808 introduced a possibility to declare fields which will be only available to parsing when a compatible API was used.

This commit replaces deprecated log with compatible logging when a 'compatible only' field was used.

relates https://github.com/elastic/elasticsearch/issues/51816

pgomulka committed:
Field parsing logging

elasticmachine commented:
Pinging @elastic/es-core-infra (Team:Core/Infra)

pgomulka committed:
add prefix

pgomulka committed:
emitt compatible warning only

pgomulka committed:
no prefix

pgomulka committed:
remove prefix from tests

pgomulka committed:
cleanup

pgomulka committed:
Merge remote-tracking branch 'upstream/master' into compat/compatible_warning_on_field

jakelandis commented on libs/x-content/src/test/java/org/elasticsearch/common/ParseFieldTests.java:
```diff
@@ -71,4 +70,22 @@ public void testGetAllNamesIncludedDeprecated() {
+
+    public void testCompatibleField() {
+        ParseField field = new ParseField("new_name", "old_name")
+            .withRestApiCompatibilityVersions(RestApiCompatibleVersion.V_7);
```
should this test use minimum version instead of `v_7` ?

jakelandis commented on libs/x-content/src/test/java/org/elasticsearch/common/ParseFieldTests.java:
```diff
@@ -71,4 +70,22 @@ public void testGetAllNamesIncludedDeprecated() {
+
+        ParseField allDepField = new ParseField("dep", "old_name")
+            .withAllDeprecated()
+            .withRestApiCompatibilityVersions(RestApiCompatibleVersion.V_7);
```
I assume this will need to be modified once `equalTo` and `onOrBefore` land ?  (can we test both cases once that lands)

jakelandis reviewed (commented):
changes are looking good .. will take another look once the merge conflicts are resolved and https://github.com/elastic/elasticsearch/pull/69774 has landed.

pgomulka committed:
Merge branch 'master' into compat/compatible_warning_on_field

pgomulka committed:
testing logging from parsefield

pgomulka committed:
Merge branch 'master' into compat/compatible_warning_on_field

joegallo commented:
I want to think a little more about what this will look like as we move from version to version, I'm not sure I like the implications of this code:

```
//a regular newly deprecated field. Emitting deprecation logs instead of compatible logs
ParseField field = new ParseField("new_name", "old_name");
```

Maybe in the presence of more examples this would be okay, but on its own I think that's going to result in a lot of work when we go from 8.x to 9.x.

I'd rather see something more along the lines of the direction I went in https://github.com/elastic/elasticsearch/pull/69573, where we'd be declaring when something was deprecated, and then emitting deprecation logs or compatible logs depending on that version.

That is, something like this, perhaps (total straw man proposal):

```
ParseField field1 = new ParseField("new_name_1", "old_name_1", RestApiVersion.V_7); 
ParseField field2 = new ParseField("new_name_2", "old_name_2", RestApiVersion.V_8);
```

Which would mean that on 8.0.0, `old_name_2` would emit deprecation warnings but that `old_name_1` would be available via `compatible-with` only, and would emit compatibility warnings.

And then the idea would be that when master is updated to be 9.0.0, we'd remove the `V_7` enum, and `field1` would be removed entirely, while  `field2` would move to being available via `compatible-with` only without needing a code change.

pgomulka commented:
I agree, we should set up time later once we have more examples and revisit this (and other infrastructure code) to make sure that it provides good experience for future developers of compatible API.
```
ParseField field = new ParseField("new_name", "old_name");
```
is just a way we deprecate fields at the moment. This has to be backed by some "mapping" code (like `request.setNewName` or sth and that code is likely to change when a deprecation is removed. For instance a deprecated field should throw error once it is removed.

```
ParseField field1 = new ParseField("new_name_1", "old_name_1", RestApiVersion.V_7); 
```
This is actually possible at the moment and I use this ins ReindexRequest
```
 PARSER.declareInt(ReindexRequest::setMaxDocsValidateIdentical,
            new ParseField("max_docs", "size")
                .forRestApiVersion(RestApiVersion.equalTo(RestApiVersion.V_7)));

        PARSER.declareInt(ReindexRequest::setMaxDocsValidateIdentical,
            new ParseField("max_docs")
                .forRestApiVersion(RestApiVersion.onOrAfter(RestApiVersion.V_8)));
```

btw this PR is purely about adding a missing compatible logging.

pgomulka committed:
missing warning assertion

pgomulka committed:
remove unsed constant

jakelandis commented on libs/x-content/src/main/java/org/elasticsearch/common/xcontent/DeprecationHandler.java:
```diff
@@ -97,4 +97,13 @@ public void usedDeprecatedField(String parserName, Supplier<XContentLocation> lo
+     * @return a variant of the current deprecation handler
+     */
+    default DeprecationHandler getInstance(boolean compatibleWarnings) {
+        return this;
```
This seems off ... such that I can pass true or false and get back the same instance. 

I would suggest to not use a default method if there are variants possible.

pgomulka replied:
I agree, that this being part of an interface is a design flaw. 
I think there is a way to choose a LoggingDeprecationHandler when creating an xcontent parser.
The RestRequest.contentParser() is aware of the version and can choose which one to use.

pgomulka replied:
actually I take that back (the restrequest part) I cannot choose the DeprecationHandler at that point, because compatible request can still access V_8 (non compatible fields, but newly deprecated) and those should not emit compatible log..

jakelandis commented on server/src/main/java/org/elasticsearch/common/xcontent/LoggingDeprecationHandler.java:
```diff
@@ -24,39 +25,60 @@
 
-    private LoggingDeprecationHandler() {
-        // Singleton
+    public static final LoggingDeprecationHandler INSTANCE = new LoggingDeprecationHandler((message, params, field_name) ->
```
Can you make these singleton instances private, so the only access is via the getter method ?

pgomulka replied:
`LoggingDeprecationHandler.INSTANCE` is used in 241 places. I am afraid that this would hide the main intent of this PR.

pgomulka replied:
also please note that this getter method is actually an instance method..

jakelandis reviewed (changes requested):
just a couple small comments

pgomulka committed:
refactor logging deprecation handler to avoid using getInstance

pgomulka committed:
javadoc cleanup

pgomulka commented:
@jakelandis I have removed the `getInstance` method and I have changed the approach. The idea now is that ParseField is using a dedicated set of methods `useDeprecatedField/Name` with an additional boolean parameter indicating if the compatible log should be used.
These methods are implemented in loggingDeprecationHandler only. Other Deprecationhandlers are not implementing this (default implementation of the new methods with boolean is delegating to old methods).

jakelandis commented on libs/x-content/src/main/java/org/elasticsearch/common/xcontent/DeprecationHandler.java:
```diff
@@ -93,8 +93,36 @@ public void usedDeprecatedField(String parserName, Supplier<XContentLocation> lo
+     * @see DeprecationHandler#usedDeprecatedName(String, Supplier, String, String)
+     * Emits a compatible api warning instead of deprecation warning when isCompatibleDeprecation is true
+     */
+    default void usedDeprecatedName(String parserName, Supplier<XContentLocation> location, String usedName, String modernName,
```
The naming here is a bit confusing.  Can we keep `useDeprecreatedField` and add in `useReplacedField` or someother way besides Name vs. Field to make this a bit more clear ?

pgomulka replied:
how about `usedRemovedField` , `usedRenamedField` and `usedReplacedField` ?

jakelandis replied:
sorry for being so pedantic... but why `used` to prefix ? would something like this read better ?

LoggingDeprecationHandler.INSTANCE.logRenamedField(null, () -> null, Request.ALLOW_NO_JOBS, Request.ALLOW_NO_MATCH);


Also can the variable/parameter names mirror the method names:

removed : `removedName`
renamed : `oldName` and `currentName`
replaced:  `replacedName`

pgomulka replied:
yes, naming is hard :) I agree log* style methods look better for the LoggingDeprecationHandler

pgomulka committed:
test ignore and methods rename

pgomulka committed:
Merge remote-tracking branch 'upstream/master' into compat/compatible_warning_on_field

pgomulka committed:
remove the ignore

pgomulka committed:
Merge remote-tracking branch 'upstream/master' into compat/compatible_warning_on_field

pgomulka commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into compat/compatible_warning_on_field

pgomulka committed:
rename

pgomulka committed:
Merge branch 'compat/compatible_warning_on_field' of github.com:pgomulka/elasticsearch into compat/compatible_warning_on_field

pgomulka committed:
parameter names

jakelandis reviewed (approved):
LGTM , thanks for the iterations

pgomulka committed:
Merge branch 'master' into compat/compatible_warning_on_field

pgomulka committed:
line length
